### PR TITLE
refactor: process pegouts before building block header

### PR DESCRIPTION
**Pull request by scottmillner**

### CODE REVIEW

File: crates/consensus/authority/src/block_builder.rs [hunk 1]
```diff
@@ -189,6 +184,90 @@ where
             }
         }
 
+        // If end of epoch, process pegouts
+        let mut witness_data: Option<Vec<Witness>> = None;
+        if is_epoch_end(block.header.number) {
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_builder.rs]: can we just use the function on the header `is_epoch?`

File: crates/consensus/authority/src/block_builder.rs [hunk 2]
```diff
@@ -189,6 +184,90 @@ where
...
+                    return;
+                }
+            };
+
+            // add current block pegouts
+            pegouts.extend(current_block_pegouts);
+
+            // send pegouts
+            info!(target: "consensus::authority", "Sending pegouts: {:?}", pegouts);
+            if let Err(e) = send_pegouts(
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_builder.rs:208]: What does send_pegouts do? Does this initiate frost signing somewhere?

File: crates/consensus/authority/src/block_builder.rs [hunk 3]
```diff
@@ -189,6 +184,90 @@ where
...
+                // Exponential backoff
+                delay *= 2;
+                tokio::time::sleep(delay).await;
+            }
+
+            // TODO(scott): check psbt matches expected session id
+
+            // extract witness data from psbt to be included in the header
+            if let Some(psbt) = psbt_message {
+                let psbt_result = Psbt::deserialize(&psbt.psbt).expect("valid psbt");
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_builder.rs]: Any reason why we shouldn't map_err here and error out?

**Review by arminsabouri — COMMENTED** [review]
Just had some questions

File: crates/consensus/authority/src/block_builder.rs (hunk 2 above)
Comment (arminsabouri) [reply crates/consensus/authority/src/block_builder.rs:208]: Calls intiate frost 
closing

File: crates/consensus/authority/src/block_builder.rs [hunk 4]
```diff
@@ -189,6 +186,73 @@ where
...
+            info!(target: "consensus::authority", "Sending pegouts: {:?}", pegouts);
+            if let Err(e) = send_pegouts(
+                &self.bitcoind_client,
+                &mut self.btc_server,
+                &self.frost_handle,
+                pegouts,
+            )
+            .await
+            {
+                error!(target: "consensus::authority", ?e, "Failed to send pegouts");
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_builder.rs:216]: Should we error out here? or continue to produce a block?

**Review by arminsabouri — APPROVED** [review]
Looks good. I would also have @BatiGencho take a look before merging

File: crates/consensus/authority/src/block_builder.rs (hunk 4 above)
Comment (scottmillner) [reply crates/consensus/authority/src/block_builder.rs:216]: ive added a `return;`
- yeah i think we should just start the loop again and try and process the pegouts again
- if they keep failing, then node will go out of turn and next member will try and seal the epoch

File: crates/consensus/authority/src/block_builder.rs (hunk 3 above)
Comment (scottmillner) [reply crates/consensus/authority/src/block_builder.rs]: refactored: error out now

**Review by BatiGencho — APPROVED** [review]

**Comment by BatiGencho** [comment]
> Looks good. I would also have @BatiGencho take a look before merging

Approved

**Pull request merged**